### PR TITLE
Move literal serialization to JPQLTemplates

### LINKED ISSUE #1205: Support Date Time literals serialization in case constructs
I'm trying to create a case expression like (simplified):
```java
new CaseBuilder()
	.when(medicalCheck.lastCheck.isNull()).then(MedicalCheckStatus.EXPIRED.ordinal())
	.when(medicalCheck.lastCheck.gt(lastCheckSoftLimit)).then(MedicalCheckStatus.VALID.ordinal())
	.otherwise(MedicalCheckStatus.PENDING.ordinal());
```

```java
import org.joda.time.LocalDate;

public class MedicalCheck implements Serializable {
	@Column(name = "last_check")
	private LocalDate lastCheck;
```

When executing this with JPA (Querydsl-3.6.1, Hibernate-4.3.8), I get this error:
```
java.lang.IllegalArgumentException: Unsupported constant 2014-02-22
	at com.mysema.query.jpa.JPQLSerializer.visitLiteral(JPQLSerializer.java:334)
	at com.mysema.query.jpa.JPQLSerializer.visitConstant(JPQLSerializer.java:298)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:208)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:42)
	at com.mysema.query.types.ConstantImpl.accept(ConstantImpl.java:127)
	at com.mysema.query.support.SerializerBase.handle(SerializerBase.java:105)
	at com.mysema.query.support.SerializerBase.visitOperation(SerializerBase.java:283)
	at com.mysema.query.jpa.JPQLSerializer.visitOperation(JPQLSerializer.java:440)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:249)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:42)
	at com.mysema.query.types.OperationImpl.accept(OperationImpl.java:90)
	at com.mysema.query.support.SerializerBase.handle(SerializerBase.java:105)
	at com.mysema.query.support.SerializerBase.visitOperation(SerializerBase.java:278)
	at com.mysema.query.jpa.JPQLSerializer.visitOperation(JPQLSerializer.java:443)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:249)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:42)
	at com.mysema.query.types.OperationImpl.accept(OperationImpl.java:90)
	at com.mysema.query.support.SerializerBase.handle(SerializerBase.java:105)
	at com.mysema.query.support.SerializerBase.visitOperation(SerializerBase.java:280)
	at com.mysema.query.jpa.JPQLSerializer.visitOperation(JPQLSerializer.java:443)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:249)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:42)
	at com.mysema.query.types.OperationImpl.accept(OperationImpl.java:90)
	at com.mysema.query.support.SerializerBase.handle(SerializerBase.java:105)
	at com.mysema.query.support.SerializerBase.visitOperation(SerializerBase.java:280)
	at com.mysema.query.jpa.JPQLSerializer.visitOperation(JPQLSerializer.java:443)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:249)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:42)
	at com.mysema.query.types.OperationImpl.accept(OperationImpl.java:90)
	at com.mysema.query.support.SerializerBase.handle(SerializerBase.java:105)
	at com.mysema.query.support.SerializerBase.visitOperation(SerializerBase.java:278)
	at com.mysema.query.jpa.JPQLSerializer.visitOperation(JPQLSerializer.java:443)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:249)
	at com.mysema.query.support.SerializerBase.visit(SerializerBase.java:42)
	at com.mysema.query.types.OperationImpl.accept(OperationImpl.java:90)
	at com.mysema.query.support.SerializerBase.handle(SerializerBase.java:105)
	at com.mysema.query.jpa.JPQLSerializer.serialize(JPQLSerializer.java:210)
	at com.mysema.query.jpa.JPAQueryBase.serialize(JPAQueryBase.java:57)
	at com.mysema.query.jpa.impl.AbstractJPAQuery.createQuery(AbstractJPAQuery.java:124)
	at com.mysema.query.jpa.impl.AbstractJPAQuery.count(AbstractJPAQuery.java:79)
	at org.springframework.data.jpa.repository.support.QueryDslJpaRepository.findAll(QueryDslJpaRepository.java:125)
```
I see that 41b73ea8166ea21e7063e8cb1dcc98c97b43dffd added a TODO for date times. Can you please complete this task? :)

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1205

### CODE REVIEW

File: querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java [hunk 1]
```diff
@@ -298,6 +308,7 @@ public void As() {
     @Test
     @NoBatooJPA
     public void Case() {
+<<<<<<< HEAD
```
Comment (Shredder121) [inline querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java]: You committed merge markers

File: querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java [hunk 2]
```diff
@@ -346,6 +402,18 @@ public void Case5() {
         assertEquals(ImmutableList.of(0, 1, 1, 1),
                 query().from(cat).orderBy(cat.id.asc())
                        .select(cat.mate.when(savedCats.get(0)).then(0).otherwise(1)).fetch());
+
+        List<Integer> rv = query().from(cat)
+            .select(Expressions.cases()
+                    .when(cat.toes.in(2, 3)).then(cat.id.multiply(cat.toes))
+                    .otherwise(4)).fetch();
+        assertInstancesOf(Integer.class, rv);
```
Comment (Shredder121) [inline querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java]: This duplicates Case3 right? It [came from there](https://github.com/querydsl/querydsl/commit/9f5b019c223e502ac10f92e22c1b08c1e4d5a160#diff-7d971dd1c9376907a88a32f2d2c5881cR370).
I think that you could move this assertion there, the projection is already there.

File: querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java [hunk 3]
```diff
@@ -318,6 +368,12 @@ public void Case2() {
                         .select(Expressions.cases().when(cat.toes.eq(2)).then(cat.id.multiply(2))
                                 .when(cat.toes.eq(3)).then(cat.id.multiply(3))
                                 .otherwise(4)).fetch());
+
+        List<Integer> rv = query().from(cat)
+            .select(Expressions.cases().when(cat.toes.eq(2)).then(cat.id.multiply(2))
```
Comment (Shredder121) [inline querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java]: This query is already run by the [statement above](https://github.com/querydsl/querydsl/pull/1208/files#diff-7d971dd1c9376907a88a32f2d2c5881cR368) right?

**Comment by johnktims** [comment]
Looks good to me.

**Comment by Shredder121** [comment]
Yes, looks good to me as well. I'll merge this.

**Pull request merged**